File: esj-core/src/main/java/de/bsnsoft/esj/json/EsjWriter.java
package de.bsnsoft.esj.json;

import de.bsnsoft.esj.Esj;
import de.bsnsoft.esj.EsjFormatException;
import de.bsnsoft.esj.ExtensionValue;
import de.bsnsoft.esj.SemanticDocument;
import de.bsnsoft.esj.SemanticPath;
import de.bsnsoft.esj.SemanticValue;
import de.bsnsoft.esj.validate.FindingCode;
import java.io.IOException;
import java.io.OutputStream;
import java.io.UncheckedIOException;
import java.io.Writer;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Iterator;
import java.util.Map;
import java.util.Objects;

/**
 * Writes a {@link SemanticDocument} as JSON, in the canonical form of the specification,
 * section 7 or in the pretty form of section 7.7.
 *
 * <p>Both forms carry the same content in the same order; they differ only in layout. The
 * canonical form is UTF-8 without a byte order mark, without insignificant whitespace and
 * without a trailing newline, and it is the only form the digests of section 8 are taken
 * over. The pretty form indents by two spaces, writes one member and one array element
 * per line, ends every line with LF and closes the file with one.
 *
 * <p>The escaping is this implementation's own (specification, section 7.5). No JSON
 * library decides it, because a library's defaults — an escaped solidus, an escaped
 * non-ASCII code point — would change the bytes and therefore both digests.
 *
 * <p>A JSON number inside {@code extensions} is written in the canonical decimal form of
 * section 7.6, rule 2 in both forms. The pretty form is allowed to keep the spelling it
 * read; this writer does not keep it, because a pretty file that is reproducible from the
 * content alone is worth more than one that remembers its input.
 */
public final class EsjWriter {

    private static final int INITIAL_SIZE = 4096;

    private final Style style;

    private EsjWriter(Style style) {
        this.style = style;
    }

    /**
     * Returns a writer for the canonical form of the specification, section 7.
     *
     * @return the canonical writer
     */
    public static EsjWriter canonical() {
        return new EsjWriter(Style.CANONICAL);
    }

    /**
     * Returns a writer for the pretty form of the specification, section 7.7.
     *
     * @return the pretty writer
     */
    public static EsjWriter pretty() {
        return new EsjWriter(Style.PRETTY);
    }

    /**
     * Returns a writer for one of the two forms.
     *
     * @param style the form to write
     * @return the writer
     * @throws NullPointerException if {@code style} is {@code null}
     */
    public static EsjWriter of(Style style) {
        return new EsjWriter(Objects.requireNonNull(style, "style"));
    }

    /**
     * Returns the form this writer writes.
     *
     * @return the style
     */
    public Style style() {
        return style;
    }

    /**
     * Writes a document to a byte stream. The stream is not closed.
     *
     * @param document the document to write
     * @param out      the stream to write to
     * @throws EsjFormatException   if a string of the document carries an unpaired
     *                              surrogate, or a number inside {@code extensions} has no
     *                              canonical decimal form of at most 64 characters
     * @throws UncheckedIOException if the stream fails
     * @throws NullPointerException if an argument is {@code null}
     */
    public void write(SemanticDocument document, OutputStream out) {
        Objects.requireNonNull(out, "out");
        try {
            out.write(toBytes(document));
        } catch (IOException e) {
            throw new UncheckedIOException("writing the document failed", e);
        }
    }

    /**
     * Writes a document to a character stream. The stream is not closed. The bytes of the
     * specification are produced first and then decoded, so that the result is the same
     * text either way; a caller that needs the exact bytes uses
     * {@link #write(SemanticDocument, OutputStream)} or {@link #toBytes(SemanticDocument)}.
     *
     * @param document the document to write
     * @param out      the stream to write to
     * @throws EsjFormatException   if a string of the document carries an unpaired
     *                              surrogate, or a number inside {@code extensions} has no
     *                              canonical decimal form of at most 64 characters
     * @throws UncheckedIOException if the stream fails
     * @throws NullPointerException if an argument is {@code null}
     */
    public void write(SemanticDocument document, Writer out) {
        Objects.requireNonNull(out, "out");
        try {
            out.write(toText(document));
        } catch (IOException e) {
            throw new UncheckedIOException("writing the document failed", e);
        }
    }

    /**
     * Returns the document as bytes.
     *
     * @param document the document to write
     * @return the UTF-8 bytes of the serialization, without a byte order mark
     * @throws EsjFormatException   if a string of the document carries an unpaired
     *                              surrogate, or a number inside {@code extensions} has no
     *                              canonical decimal form of at most 64 characters
     * @throws NullPointerException if {@code document} is {@code null}
     */
    public byte[] toBytes(SemanticDocument document) {
        Objects.requireNonNull(document, "document");
        JsonBytes out = new JsonBytes(INITIAL_SIZE);
        writeDocument(document, out);
        return out.toByteArray();
    }

    /**
     * Returns the document as text.
     *
     * @param document the document to write
     * @return the serialization as a string
     * @throws EsjFormatException   if a string of the document carries an unpaired
     *                              surrogate, or a number inside {@code extensions} has no
     *                              canonical decimal form of at most 64 characters
     * @throws NullPointerException if {@code document} is {@code null}
     */
    public String toText(SemanticDocument document) {
        return new String(toBytes(document), StandardCharsets.UTF_8);
    }

    /**
     * Returns the semantic identity of a document, serialized as the standalone
     * two-member object of the specification, section 8.2: the edition and the values, in
     * that order. It is the byte sequence the semantic digest is taken over.
     *
     * <p>The edition is inside it because a path is an address relative to an edition: two
     * documents carrying the same strings under two editions would otherwise collide in
     * one digest, and a deduplication key built on it would be silently wrong for exactly
     * the documents that straddle an edition change.
     *
     * @param document the document to write
     * @return the bytes of the semantic identity object
     */
    byte[] semanticToBytes(SemanticDocument document) {
        Objects.requireNonNull(document, "document");
        JsonBytes out = new JsonBytes(INITIAL_SIZE);
        out.ascii('{');
        member(out, 1, true, "semanticModel");
        out.string(document.semanticModel());
        member(out, 1, false, "values");
        writeValues(document, out, 1);
        close(out, 0, '}');
        return out.toByteArray();
    }

    private void writeDocument(SemanticDocument document, JsonBytes out) {
        out.ascii('{');
        member(out, 1, true, "format");
        out.string(Esj.FORMAT);
        member(out, 1, false, "version");
        out.string(Esj.VERSION);
        member(out, 1, false, "semanticModel");
        out.string(document.semanticModel());
        member(out, 1, false, "values");
        writeValues(document, out, 1);
        if (!document.extensions().isEmpty()) {
            member(out, 1, false, "extensions");
            writeExtensions(document.extensions(), out, 1);
        }
        document.source().ifPresent(source -> {
            member(out, 1, false, "source");
            writeSource(source, out, 1);
        });
        close(out, 0, '}');
        if (style == Style.PRETTY) {
            out.ascii('\n');
        }
    }

    private void writeValues(SemanticDocument document, JsonBytes out, int depth) {
        if (document.values().isEmpty()) {
            out.ascii("{}");
            return;
        }
        out.ascii('{');
        boolean first = true;
        for (Map.Entry<SemanticPath, SemanticValue> entry : document.values().entrySet()) {
            member(out, depth + 1, first, entry.getKey().toString());
            writeValue(entry.getValue(), out, depth + 1);
            first = false;
        }
        close(out, depth, '}');
    }

    /**
     * Writes one value. A value that carries no supplementary component is a JSON string
     * and a value that carries one is a JSON object, with the members in the order of the
     * specification, section 7.3, rule 3. The content is written as it stands: this writer
     * serializes a document and never rewrites what a value holds (section 6.4).
     */
    private void writeValue(SemanticValue value, JsonBytes out, int depth) {
        if (!value.hasComponents()) {
            out.string(value.canonicalContent());
            return;
        }
        out.ascii('{');
        member(out, depth + 1, true, "value");
        out.string(value.canonicalContent());
        component(out, depth, "scheme", value.scheme());
        component(out, depth, "schemeVersion", value.schemeVersion());
        component(out, depth, "mimeCode", value.mimeCode());
        component(out, depth, "filename", value.filename());
        close(out, depth, '}');
    }

    private void component(JsonBytes out, int depth, String name, String value) {
        if (value != null) {
            member(out, depth + 1, false, name);
            out.string(value);
        }
    }

    private void writeSource(SemanticDocument.Source source, JsonBytes out, int depth) {
        out.ascii('{');
        boolean first = true;
        if (source.syntax().isPresent()) {
            member(out, depth + 1, true, "syntax");
            out.string(source.syntax().orElseThrow());
            first = false;
        }
        if (source.sha256().isPresent()) {
            member(out, depth + 1, first, "sha256");
            out.string(source.sha256().orElseThrow());
        }
        close(out, depth, '}');
    }

    private void writeExtensions(Map<String, ExtensionValue> extensions, JsonBytes out, int depth) {
        out.ascii('{');
        boolean first = true;
        for (Map.Entry<String, ExtensionValue> entry : extensions.entrySet()) {
            member(out, depth + 1, first, entry.getKey());
            writeExtensionValue(entry.getValue(), out, depth + 1);
            first = false;
        }
        close(out, depth, '}');
    }

    /**
     * Writes one extension subtree. The walk is iterative: a document assembled through
     * the public API may nest deeper than a document the reader would have accepted
     * (specification, section 12.2), and a recursive writer would answer that with a
     * {@link StackOverflowError} rather than with a serialization.
     */
    private void writeExtensionValue(ExtensionValue value, JsonBytes out, int depth) {
        Deque<Frame> open = new ArrayDeque<>();
        ExtensionValue pending = value;
        int pendingDepth = depth;
        while (true) {
            Frame opened = writeNode(pending, out, pendingDepth);
            if (opened != null) {
                open.push(opened);
            }
            pending = null;
            while (pending == null) {
                if (open.isEmpty()) {
                    return;
                }
                Frame frame = open.peek();
                if (!frame.hasNext()) {
                    close(out, frame.depth(), frame.bracket());
                    open.pop();
                    continue;
                }
                pendingDepth = frame.depth() + 1;
                pending = frame.next(out);
            }
        }
    }

    /**
     * Writes a node and returns the frame of a container that has children, or
     * {@code null} for a scalar and for an empty container.
     */
    private Frame writeNode(ExtensionValue value, JsonBytes out, int depth) {
        if (value instanceof ExtensionValue.ObjectValue object) {
            if (object.members().isEmpty()) {
                out.ascii("{}");
                return null;
            }
            out.ascii('{');
            return Frame.of(object.members().entrySet().iterator(), null, '}', depth, this);
        }
        if (value instanceof ExtensionValue.ArrayValue array) {
            if (array.elements().isEmpty()) {
                out.ascii("[]");
                return null;
            }
            out.ascii('[');
            return Frame.of(null, array.elements().iterator(), ']', depth, this);
        }
        if (value instanceof ExtensionValue.StringValue string) {
            out.string(string.value());
        } else if (value instanceof ExtensionValue.NumberValue number) {
            out.ascii(canonicalNumber(number.value()));
        } else if (value instanceof ExtensionValue.BooleanValue bool) {
            out.ascii(bool.value() ? "true" : "false");
        } else {
            out.ascii("null");
        }
        return null;
    }

    /** One container the writer has opened and not yet closed. */
    private static final class Frame {

        private final Iterator<Map.Entry<String, ExtensionValue>> members;
        private final Iterator<ExtensionValue> elements;
        private final char bracket;
        private final int depth;
        private final EsjWriter writer;
        private boolean first = true;

        private Frame(Iterator<Map.Entry<String, ExtensionValue>> members,
                      Iterator<ExtensionValue> elements,
                      char bracket,
                      int depth,
                      EsjWriter writer) {
            this.members = members;
            this.elements = elements;
            this.bracket = bracket;
            this.depth = depth;
            this.writer = writer;
        }

        static Frame of(Iterator<Map.Entry<String, ExtensionValue>> members,
                        Iterator<ExtensionValue> elements,
                        char bracket,
                        int depth,
                        EsjWriter writer) {
            return new Frame(members, elements, bracket, depth, writer);
        }

        char bracket() {
            return bracket;
        }

        int depth() {
            return depth;
        }

        boolean hasNext() {
            return members != null ? members.hasNext() : elements.hasNext();
        }

        /** Writes the separator and the member name, and returns the child to write. */
        ExtensionValue next(JsonBytes out) {
            ExtensionValue value;
            if (members != null) {
                Map.Entry<String, ExtensionValue> entry = members.next();
                writer.member(out, depth + 1, first, entry.getKey());
                value = entry.getValue();
            } else {
                writer.element(out, depth + 1, first);
                value = elements.next();
            }
            first = false;
            return value;
        }
    }

    private static String canonicalNumber(BigDecimal value) {
        String canonical = Decimals.canonicalize(value.toString());
        if (canonical == null) {
            throw new EsjFormatException(
                    "a number inside extensions has a canonical decimal form of more than "
                            + Decimals.MAX_LENGTH + " characters",
                    FindingCode.ESJ_L1_EXT_NUMBER, null);
        }
        return canonical;
    }

    private void member(JsonBytes out, int depth, boolean first, String name) {
        element(out, depth, first);
        out.string(name);
        out.ascii(':');
        if (style == Style.PRETTY) {
            out.ascii(' ');
        }
    }

    private void element(JsonBytes out, int depth, boolean first) {
        if (!first) {
            out.ascii(',');
        }
        indent(out, depth);
    }

    private void close(JsonBytes out, int depth, char bracket) {
        indent(out, depth);
        out.ascii(bracket);
    }

    private void indent(JsonBytes out, int depth) {
        if (style != Style.PRETTY) {
            return;
        }
        out.ascii('\n');
        for (int i = 0; i < depth; i++) {
            out.ascii("  ");
        }
    }

    /** The two serializations the specification defines. */
    public enum Style {

        /**
         * The canonical form of the specification, section 7: one byte sequence per
         * document content, with no insignificant whitespace and no trailing newline.
         */
        CANONICAL,

        /**
         * The pretty form of the specification, section 7.7: the same content and the same
         * order, indented by two spaces, one member per line, LF line endings and a
         * trailing newline.
         */
        PRETTY
    }
}

File: bindings/csharp/En16931.SemanticJson/Json/EsjWriter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace En16931.SemanticJson.Json;

/// <summary>
/// Writes a document as JSON, in the canonical form of the specification, section 7 or in
/// the pretty form of section 7.7.
/// </summary>
/// <remarks>
/// Every byte is decided here rather than by a JSON library, because a library's defaults —
/// an escaped solidus, an escaped non-ASCII character, a space after a colon — would each
/// change the canonical form and with it both digests. The content of a value is written as
/// it stands: this writer serializes a document and never rewrites what a value holds. A
/// JSON number inside <c>extensions</c> is written in the canonical decimal form of
/// section 7.6, rule 2, which is string processing over the digits the document carried.
/// </remarks>
public sealed class EsjWriter
{
    private readonly bool _pretty;

    private EsjWriter(bool pretty)
    {
        _pretty = pretty;
    }

    /// <summary>Returns a writer for the canonical form of the specification, section 7.</summary>
    /// <returns>the canonical writer</returns>
    public static EsjWriter Canonical() => new(false);

    /// <summary>Returns a writer for the pretty form of the specification, section 7.7.</summary>
    /// <returns>the pretty writer</returns>
    public static EsjWriter Pretty() => new(true);

    /// <summary>Writes a document and returns its bytes.</summary>
    /// <param name="document">the document</param>
    /// <returns>the bytes, encoded in UTF-8 without a byte order mark</returns>
    public byte[] ToBytes(SemanticDocument document)
    {
        ArgumentNullException.ThrowIfNull(document);
        Bytes bytes = new();
        WriteDocument(document, bytes);
        return bytes.ToArray();
    }

    /// <summary>Writes a document and returns it as text.</summary>
    /// <param name="document">the document</param>
    /// <returns>the JSON text</returns>
    public string ToText(SemanticDocument document) =>
        new UTF8Encoding(false).GetString(ToBytes(document));

    /// <summary>Writes a document to a stream.</summary>
    /// <param name="document">the document</param>
    /// <param name="output">the stream, which is written to and not closed</param>
    public void Write(SemanticDocument document, Stream output)
    {
        ArgumentNullException.ThrowIfNull(output);
        byte[] bytes = ToBytes(document);
        output.Write(bytes, 0, bytes.Length);
    }

    /// <summary>
    /// Writes the object the semantic digest is taken over: the edition and the values, and
    /// nothing else (specification, section 8.2).
    /// </summary>
    /// <param name="document">the document</param>
    /// <returns>the canonical bytes of that object</returns>
    public byte[] SemanticToBytes(SemanticDocument document)
    {
        ArgumentNullException.ThrowIfNull(document);
        Bytes bytes = new();
        bytes.Ascii('{');
        WriteMember(bytes, 1, true, "semanticModel");
        bytes.String(document.SemanticModel);
        WriteMember(bytes, 1, false, "values");
        WriteValues(document, bytes, 1);
        Close(bytes, 0, '}');
        return bytes.ToArray();
    }

    private void WriteDocument(SemanticDocument document, Bytes bytes)
    {
        bytes.Ascii('{');
        WriteMember(bytes, 1, true, "format");
        bytes.String(Esj.Format);
        WriteMember(bytes, 1, false, "version");
        bytes.String(Esj.Version);
        WriteMember(bytes, 1, false, "semanticModel");
        bytes.String(document.SemanticModel);
        WriteMember(bytes, 1, false, "values");
        WriteValues(document, bytes, 1);
        if (document.Extensions.Count > 0)
        {
            WriteMember(bytes, 1, false, "extensions");
            WriteExtensions(document.Extensions, bytes, 1);
        }

        if (document.Source is not null)
        {
            WriteMember(bytes, 1, false, "source");
            WriteSource(document.Source, bytes, 1);
        }

        Close(bytes, 0, '}');
        if (_pretty)
        {
            bytes.Ascii('\n');
        }
    }

    private void WriteValues(SemanticDocument document, Bytes bytes, int depth)
    {
        if (document.Values.Count == 0)
        {
            bytes.Ascii("{}");
            return;
        }

        bytes.Ascii('{');
        bool first = true;
        foreach (KeyValuePair<SemanticPath, SemanticValue> entry in document.Entries)
        {
            WriteMember(bytes, depth + 1, first, entry.Key.Text);
            WriteValue(entry.Value, bytes, depth + 1);
            first = false;
        }

        Close(bytes, depth, '}');
    }

    private void WriteValue(SemanticValue value, Bytes bytes, int depth)
    {
        if (!value.HasComponent)
        {
            bytes.String(value.Content);
            return;
        }

        bytes.Ascii('{');
        WriteMember(bytes, depth + 1, true, "value");
        bytes.String(value.Content);
        Component(bytes, depth, "scheme", value.Scheme);
        Component(bytes, depth, "schemeVersion", value.SchemeVersion);
        Component(bytes, depth, "mimeCode", value.MimeCode);
        Component(bytes, depth, "filename", value.Filename);
        Close(bytes, depth, '}');
    }

    private void Component(Bytes bytes, int depth, string name, string? value)
    {
        if (value is not null)
        {
            WriteMember(bytes, depth + 1, false, name);
            bytes.String(value);
        }
    }

    private void WriteSource(DocumentSource source, Bytes bytes, int depth)
    {
        bytes.Ascii('{');
        bool first = true;
        if (source.Syntax is not null)
        {
            WriteMember(bytes, depth + 1, true, "syntax");
            bytes.String(source.Syntax);
            first = false;
        }

        if (source.Sha256 is not null)
        {
            WriteMember(bytes, depth + 1, first, "sha256");
            bytes.String(source.Sha256);
        }

        Close(bytes, depth, '}');
    }

    private void WriteExtensions(
        IReadOnlyList<KeyValuePair<string, ExtensionValue>> extensions, Bytes bytes, int depth)
    {
        bytes.Ascii('{');
        bool first = true;
        foreach (KeyValuePair<string, ExtensionValue> entry in extensions)
        {
            WriteMember(bytes, depth + 1, first, entry.Key);
            WriteExtensionValue(entry.Value, bytes, depth + 1);
            first = false;
        }

        Close(bytes, depth, '}');
    }

    /// <summary>
    /// Writes one extension subtree. The walk is iterative, because a document assembled
    /// through this API may nest deeper than a document the reader would have accepted, and a
    /// recursive writer would answer that with a stack overflow rather than a serialization.
    /// </summary>
    private void WriteExtensionValue(ExtensionValue value, Bytes bytes, int depth)
    {
        Stack<Frame> open = new();
        ExtensionValue? pending = value;
        int pendingDepth = depth;
        while (true)
        {
            Frame? frame = WriteNode(pending!, bytes, pendingDepth);
            if (frame is not null)
            {
                open.Push(frame);
            }

            pending = null;
            while (pending is null)
            {
                if (open.Count == 0)
                {
                    return;
                }

                Frame top = open.Peek();
                if (!top.HasNext)
                {
                    Close(bytes, top.Depth, top.Bracket);
                    open.Pop();
                    continue;
                }

                pendingDepth = top.Depth + 1;
                pending = top.Next(this, bytes);
            }
        }
    }

    private Frame? WriteNode(ExtensionValue value, Bytes bytes, int depth)
    {
        switch (value.Kind)
        {
            case ExtensionKind.Object:
                if (value.Members.Count == 0)
                {
                    bytes.Ascii("{}");
                    return null;
                }

                bytes.Ascii('{');
                return Frame.OfObject(value.Members, depth);
            case ExtensionKind.Array:
                if (value.Elements.Count == 0)
                {
                    bytes.Ascii("[]");
                    return null;
                }

                bytes.Ascii('[');
                return Frame.OfArray(value.Elements, depth);
            case ExtensionKind.String:
                bytes.String(value.Text);
                return null;
            case ExtensionKind.Number:
                bytes.Ascii(CanonicalNumber(value.Text));
                return null;
            case ExtensionKind.Boolean:
                bytes.Ascii(value.Text);
                return null;
            default:
                bytes.Ascii("null");
                return null;
        }
    }

    private static string CanonicalNumber(string lexical)
    {
        string? canonical = Decimals.Canonicalize(lexical);
        return canonical ?? throw new EsjFormatException("ESJ-L1-EXT-NUMBER",
            "a number inside extensions has a canonical decimal form of more than "
            + Decimals.MaxLength + " characters");
    }

    private void WriteMember(Bytes bytes, int depth, bool first, string name)
    {
        if (!first)
        {
            bytes.Ascii(',');
        }

        Indent(bytes, depth);
        bytes.String(name);
        bytes.Ascii(':');
        if (_pretty)
        {
            bytes.Ascii(' ');
        }
    }

    private void Element(Bytes bytes, int depth, bool first)
    {
        if (!first)
        {
            bytes.Ascii(',');
        }

        Indent(bytes, depth);
    }

    private void Close(Bytes bytes, int depth, char bracket)
    {
        Indent(bytes, depth);
        bytes.Ascii(bracket);
    }

    private void Indent(Bytes bytes, int depth)
    {
        if (!_pretty)
        {
            return;
        }

        bytes.Ascii('\n');
        for (int level = 0; level < depth; level++)
        {
            bytes.Ascii("  ");
        }
    }

    /// <summary>One container the writer has opened and not yet closed.</summary>
    private sealed class Frame
    {
        private readonly IReadOnlyList<KeyValuePair<string, ExtensionValue>>? _members;
        private readonly IReadOnlyList<ExtensionValue>? _elements;
        private int _at;

        private Frame(
            IReadOnlyList<KeyValuePair<string, ExtensionValue>>? members,
            IReadOnlyList<ExtensionValue>? elements,
            char bracket,
            int depth)
        {
            _members = members;
            _elements = elements;
            Bracket = bracket;
            Depth = depth;
        }

        internal char Bracket { get; }

        internal int Depth { get; }

        internal bool HasNext => _at < (_members?.Count ?? _elements!.Count);

        internal static Frame OfObject(IReadOnlyList<KeyValuePair<string, ExtensionValue>> members, int depth) =>
            new(members, null, '}', depth);

        internal static Frame OfArray(IReadOnlyList<ExtensionValue> elements, int depth) =>
            new(null, elements, ']', depth);

        internal ExtensionValue Next(EsjWriter writer, Bytes bytes)
        {
            bool first = _at == 0;
            if (_members is not null)
            {
                KeyValuePair<string, ExtensionValue> member = _members[_at++];
                writer.WriteMember(bytes, Depth + 1, first, member.Key);
                return member.Value;
            }

            ExtensionValue element = _elements![_at++];
            writer.Element(bytes, Depth + 1, first);
            return element;
        }
    }

    /// <summary>
    /// The bytes of a JSON text, written one at a time: the escaping of the specification,
    /// section 7.5 and the UTF-8 encoding are this class and no library's.
    /// </summary>
    private sealed class Bytes
    {
        private const string Hex = "0123456789abcdef";

        private readonly MemoryStream _out = new(4096);

        internal void Ascii(char character) => _out.WriteByte((byte)character);

        internal void Ascii(string literal)
        {
            foreach (char character in literal)
            {
                _out.WriteByte((byte)character);
            }
        }

        internal void String(string value)
        {
            _out.WriteByte((byte)'"');
            for (int at = 0; at < value.Length; at++)
            {
                char character = value[at];
                switch (character)
                {
                    case '"':
                        Ascii("\\\"");
                        break;
                    case '\\':
                        Ascii("\\\\");
                        break;
                    case '\b':
                        Ascii("\\b");
                        break;
                    case '\t':
                        Ascii("\\t");
                        break;
                    case '\n':
                        Ascii("\\n");
                        break;
                    case '\f':
                        Ascii("\\f");
                        break;
                    case '\r':
                        Ascii("\\r");
                        break;
                    default:
                        WriteCharacter(value, character, ref at);
                        break;
                }
            }

            _out.WriteByte((byte)'"');
        }

        internal byte[] ToArray() => _out.ToArray();

        private void WriteCharacter(string value, char character, ref int at)
        {
            if (character < 0x20)
            {
                Ascii("\\u00");
                _out.WriteByte((byte)Hex[(character >> 4) & 0xF]);
                _out.WriteByte((byte)Hex[character & 0xF]);
                return;
            }

            if (character < 0x80)
            {
                _out.WriteByte((byte)character);
                return;
            }

            if (character < 0x800)
            {
                _out.WriteByte((byte)(0xC0 | (character >> 6)));
                _out.WriteByte((byte)(0x80 | (character & 0x3F)));
                return;
            }

            if (char.IsHighSurrogate(character))
            {
                if (at + 1 >= value.Length || !char.IsLowSurrogate(value[at + 1]))
                {
                    throw LoneSurrogate();
                }

                int codePoint = char.ConvertToUtf32(character, value[++at]);
                _out.WriteByte((byte)(0xF0 | (codePoint >> 18)));
                _out.WriteByte((byte)(0x80 | ((codePoint >> 12) & 0x3F)));
                _out.WriteByte((byte)(0x80 | ((codePoint >> 6) & 0x3F)));
                _out.WriteByte((byte)(0x80 | (codePoint & 0x3F)));
                return;
            }

            if (char.IsLowSurrogate(character))
            {
                throw LoneSurrogate();
            }

            _out.WriteByte((byte)(0xE0 | (character >> 12)));
            _out.WriteByte((byte)(0x80 | ((character >> 6) & 0x3F)));
            _out.WriteByte((byte)(0x80 | (character & 0x3F)));
        }

        private static EsjFormatException LoneSurrogate() => new("ESJ-L1-SURROGATE",
            "a string carries an unpaired surrogate and therefore has no UTF-8 encoding");
    }
}
